### DISCUSSION TITLE
Potential to write EXIF without re-encoding/altering image data?

### DISCUSSION THREAD

**Discussion by dlong500** [libvips/libvips#2980] [Ideas]
See https://github.com/lovell/sharp/issues/3327

After getting confirmation from @lovell that sharp can't currently write EXIF info without re-encoding/altering JPEG image data, I'm assuming this is a limitation inherent in libvips.  Is there any possibility that support could be added in libvips to modify/replace _only_ EXIF data without touching image data?

Coming from the perspective of javascript apps, the only real alternative for actually writing EXIF data is Piexifjs which is unfortunately abandoned for almost 3 years at this point.  Is a feature like this feasible?  I see a few other issues with somewhat related ideas but no ultimate guidance on whether this is a design choice for libvips or just something that would be relatively hard to implement.  In any case I think it would be a great fit for a well-maintained library that is used as the underlying engine for a variety of other utilities.

**Comment by jcupitt** [comment]
Hi @dlong500,

As @lovell says, I think this is out of scope for libvips. libvips always decodes and reencodes. 

You need very low-level control of the file format, and that really needs something specialised. I made something to add/replace an ICC profile in a JPG a while ago:

https://gist.github.com/jcupitt/058bf71cf0de53f8b9207c221719d21c

You need that, but doing EXIF instead of ICC.

**Reply by dlong500** [reply]
Thanks.  I do understand the need for low-level control of the file format.  I might be inclined to disagree in theory with it being out of scope since modifying EXIF data is still a type of image processing.  It might be nice to have a collection of the type of utility functions such as your ICC profile example exposed within libvips if only because that would make higher-level libraries have access to such useful functions (which then makes doing such operations in other languages like javascript that much easier).

Still, I understand where you are coming from.  The issue with using something like your ICC profile code is that I would need to create a javascript wrapper in order to actually use it in a javascript app, which would essentially mean creating a library like sharp to wrap that code just to deal with EXIF writes (albeit a much smaller scale library).  From my perspective that seems like a good fit to be included in an already established image processing library that can then add value to other consumer libraries.

**Reply by jcupitt** [reply]
You can edit EXIF with libvips, you just can't avoid the decode and reencode. I think all image processing libraries are like this -- ImageMagick is the same (for example). 

If you're running node on a linux host, you could shell out to `exiftool`. That'd be very simple and perform reasonably well.
